### ISSUE TITLE
EPIC: Pitch-Level Model Pipeline with CRISP-DM Methodology [Phase 1-6 Implementation]

### ISSUE THREAD

**Issue by cbwinslow** [cbwinslow/retrosheet#78]

**Comment by cbwinslow** [comment]
## 🎯 EPIC OVERVIEW: Statcast Pitch-Level Model Pipeline

This epic implements a comprehensive, reproducible pitch-level prediction pipeline following the **CRISP-DM (Cross Industry Standard Process for Data Mining)** methodology. The goal is to transform our 7.66M pitch dataset into production-ready models.

---

## 📊 CRISP-DM PHASE BREAKDOWN

### Phase 1: Business Understanding ✅ FOUNDATION COMPLETE
**Objective:** Predict pitch outcomes using Statcast pitch-level data to enable:
- Real-time plate appearance outcome prediction
- Pitch sequencing strategy insights  
- Swing probability estimation
- Pitch quality assessment

**Success Metrics:**
- Pitch outcome classification: 60-65% accuracy (Tier 1: Ball/Strike/BIP)
- Pitch type prediction: 63-70% accuracy
- Swing probability: 76-80% accuracy
- Latency: <50ms inference time per pitch

### Phase 2: Data Understanding ✅ SUPERIOR DATA IN HAND
**Current Assets:**
- `features_pitch.locations`: 7,661,992 pitches (2015-2025)
- 118 Statcast fields including physics, location, outcome
- 100% player linkage via bridge tables
- PostGIS-enabled for spatial queries

**Data Quality:** 98.5% completeness, quality_flag classifications applied

### Phase 3: Data Preparation 🔄 FLEXIBLE SCHEMA DESIGN
**Key Design Principle:** ALL fields available in tables for selective feature engineering

**Target Schema:**
```
features_pitch.base_features       -- 118 columns, all raw fields
features_pitch.engineered_features -- derived metrics (selectable)
features_pitch.sequential_features -- pitch sequence windows
features_pitch.player_context      -- rolling player stats
features_pitch.model_training      -- training-ready merged set
```

### Phase 4: Modeling 🔄 IMPLEMENTATION PHASE
**Model Families:**
1. **Two-Tier XGBoost** (schilamkur approach) - Ball/Strike/BIP → Hit Type
2. **LSTM Sequencer** (SebastianJRamirezA approach) - Next pitch prediction
3. **Multi-Task Network** - Joint type + location prediction
4. **Swing Probability** - Binary swing/take classifier

### Phase 5: Evaluation 🔄 VALIDATION FRAMEWORK
- Season-stratified k-fold cross validation
- Out-of-time validation (train 2015-2022, test 2023-2025)
- Calibration checking (ECE < 0.02)
- Subgroup analysis by count, handedness, pitch type

### Phase 6: Deployment 🔄 PRODUCTION READY
- Model registry in PostgreSQL `models` schema
- Automated promotion pipeline
- API inference endpoints
- Real-time feature materialization

---

## 🏗️ DELIVERABLES & SUB-ISSUES

### Sub-Issue #78.1: Flexible Feature Mart Schema Design
**Branch:** `feature/pitch-mart-schema`
**Labels:** `epic-pitch`, `data-prep`, `schema-design`

Deliverables:
- [ ] ERD diagrams for all feature marts (PlantUML format)
- [ ] `features_pitch.base_features` - raw 118-field preservation
- [ ] `features_pitch.feature_registry` - metadata table for field selection
- [ ] `features_pitch.engineered_features` - derived metrics table
- [ ] `features_pitch.sequential_features` - LSTM-ready sequence tables
- [ ] SQL for flexible field inclusion/exclusion system
- [ ] Documentation: `docs/PITCH_FEATURE_MART_SCHEMA.md`

### Sub-Issue #78.2: Two-Tier XGBoost Pitch Outcome Model
**Branch:** `model/pitch-outcome-xgboost`
**Labels:** `epic-pitch`, `modeling`, `xgboost`

Deliverables:
- [ ] `scripts/models/pitch_outcome_xgboost.py` - Two-tier classifier
- [ ] Tier 1: Ball / Strike / Ball-in-Play classification
- [ ] Tier 2: Single / Double / Triple / HR / Out (if BIP)
- [ ] Feature importance analysis (SHAP values)
- [ ] Cross-validation results (season-stratified)
- [ ] Model registry entry in `models.trained_models`

### Sub-Issue #78.3: LSTM Pitch Sequence Model
**Branch:** `model/pitch-sequence-lstm`
**Labels:** `epic-pitch`, `modeling`, `lstm`, `deep-learning`

Deliverables:
- [ ] `scripts/models/pitch_sequence_lstm.py` - Sequential model
- [ ] Sequence creation: sliding windows of 3-5 pitches per at-bat
- [ ] Architecture: LSTM(128) → LSTM(64) → Dense → Output
- [ ] Predicts: pitch_type, plate_x, plate_z (next pitch)
- [ ] Training data pipeline for sequences
- [ ] Model evaluation on held-out at-bats

### Sub-Issue #78.4: Multi-Task Pitch Type + Location Model
**Branch:** `model/pitch-multitask`
**Labels:** `epic-pitch`, `modeling`, `multitask`, `research`

Deliverables:
- [ ] `scripts/models/pitch_multitask.py` - Joint prediction model
- [ ] Shared encoder → separate heads for type/location
- [ ] Multi-task loss function with weighting
- [ ] Evaluation: type accuracy + location MSE
- [ ] Comparison against separate single-task models

### Sub-Issue #78.5: Swing Probability Model
**Branch:** `model/swing-probability`
**Labels:** `epic-pitch`, `modeling`, `swing`, `lightgbm`

Deliverables:
- [ ] `scripts/models/swing_probability.py` - Binary classifier
- [ ] Target: swing (1) vs take (0)
- [ ] Separate models for In-Zone vs Out-of-Zone
- [ ] Features: pitch physics, count, game state
- [ ] Target accuracy: 80%+
- [ ] Calibration validation

### Sub-Issue #78.6: Model Registry & Evaluation Framework
**Branch:** `infra/model-registry`
**Labels:** `epic-pitch`, `infrastructure`, `evaluation`

Deliverables:
- [ ] `sql/models/001_model_registry.sql` - Registry schema
- [ ] `scripts/evaluation/cross_validate.py` - CV framework
- [ ] `scripts/evaluation/calibration_check.py` - Calibration plots
- [ ] `scripts/evaluation/shap_analysis.py` - Feature importance
- [ ] Automated model promotion pipeline
- [ ] Performance dashboard SQL views

---

## 🔄 CRISP-DM PHASES AS GITHUB PROJECT COLUMNS

This epic will be tracked on the **"Retrosheet Warehouse Roadmap"** project board with columns:

| Column | Phase | Status |
|--------|-------|--------|
| 📋 Business Understanding | Phase 1 | ✅ Complete |
| 🔍 Data Understanding | Phase 2 | ✅ Complete |
| 🔧 Data Preparation | Phase 3 | 🔄 In Progress |
| 🤖 Modeling | Phase 4 | ⏳ Planned |
| 📊 Evaluation | Phase 5 | ⏳ Planned |
| 🚀 Deployment | Phase 6 | ⏳ Planned |
| ✅ Done | All | Accumulating |

---

## 📝 METHODOLOGY & REPRODUCIBILITY

### Branch Strategy
```
main (always deployable)
  ├── develop (integration branch)
  │   ├── feature/pitch-mart-schema
  │   ├── model/pitch-outcome-xgboost
  │   ├── model/pitch-sequence-lstm
  │   ├── model/pitch-multitask
  │   ├── model/swing-probability
  │   └── infra/model-registry
  └── hotfix/* (emergency fixes)
```

### Data Lineage & Reproducibility
- All model training data queries stored in `sql/models/training_data/`
- Model artifacts versioned with data hash
- Feature engineering steps logged in `features_pitch.feature_registry`
- Every model trained can be reproduced from warehouse state

### Documentation Requirements
- [ ] `docs/PITCH_MODELING_CRISP_DM.md` - Full methodology
- [ ] `docs/PITCH_FEATURE_MART_ERD.md` - ERD diagrams
- [ ] `docs/PITCH_MODEL_COMPARISON.md` - Benchmark results
- [ ] `docs/PITCH_MODEL_API.md` - Inference API specification
- [ ] Update `AGENTS.md` with pitch modeling section
- [ ] Update `FILE_INVENTORY.md` with new scripts

---

## 🏷️ LABELS & CATEGORIZATION

**Epic Label:** `epic-pitch` (new label to create)
**Phase Labels:** `crisp-business`, `crisp-data`, `crisp-prep`, `crisp-model`, `crisp-eval`, `crisp-deploy`
**Type Labels:** `modeling`, `schema-design`, `deep-learning`, `evaluation`, `infrastructure`

---

## 🔗 RELATED ISSUES & DOCUMENTATION

- Parent Research: `docs/STATCAST_MODELS_RESEARCH_REPORT.md`
- Data Architecture: `docs/PITCH_PLAYER_ANALYSIS_ARCHITECTURE.md`
- Comparison Report: `docs/PITCH_KB_COMPARISON_REPORT.md`
- Related Epic: #63 (Modular Prediction Framework)
- Related: #30 (Live PA Feature Parity)
- Related: #31 (Live PA Outcome Scoring)

---

## 📅 TIMELINE ESTIMATES

| Phase | Sub-Issue | Duration | Target |
|-------|-----------|----------|--------|
| Data Prep | #78.1 | 1 week | Week 1 |
| Modeling | #78.2 | 1 week | Week 2 |
| Modeling | #78.3 | 2 weeks | Weeks 3-4 |
| Modeling | #78.4 | 2 weeks | Weeks 5-6 |
| Modeling | #78.5 | 1 week | Week 7 |
| Eval/Deploy | #78.6 | 1 week | Week 8 |

**Total Epic Duration:** ~8 weeks (2 months)

---

## ✅ ACCEPTANCE CRITERIA

This epic is complete when:
1. [ ] All 6 sub-issues closed and merged to `main`
2. [ ] ERD diagrams published to project documentation
3. [ ] At least 2 models achieve target accuracy metrics
4. [ ] Model registry operational with 3+ registered models
5. [ ] Evaluation framework produces automated reports
6. [ ] All documentation updated per requirements
7. [ ] CRISP-DM phases documented and linked in project board

---

**Next Action:** Create Sub-Issue #78.1 and begin Phase 3 (Data Preparation) - Flexible Schema Design

**Comment by cbwinslow** [comment]
## Progress Update: Feature Mart Schema Implementation

**Status:** 🔄 In Progress  
**Branch:** `feature/pitch-mart-schema` (created from main)

### Files Created

1. **SQL Schema** (`sql/features/003_pitch_flexible_mart.sql`)
   - `base_features` table - ALL 118 Statcast fields preserved
   - `feature_registry` table - Metadata catalog for dynamic feature selection
   - `engineered_features` table - Derived metrics (NULL-free)
   - `sequential_features` table - LSTM/Transformer sequences with JSONB
   - `player_context` table - Rolling statistics (30day/season/career)
   - `model_training_set` table - Versioned training data with data_hash
   - `pitch_sequences` table - PA-level aggregation
   - `vw_xgboost_base` view - Pre-built joined feature set
   - Helper functions for dynamic feature selection
   - Initial population of feature_registry with 30+ base features

2. **ERD Diagram** (`docs/diagrams/PITCH_FEATURE_MART_ERD.puml`)
   - Complete entity relationship diagram
   - Shows all 7 tables in the feature mart schema
   - Relationships and cardinality
   - Package grouping by schema

3. **Data Flow Diagram** (`docs/diagrams/PITCH_DATA_FLOW.puml`)
   - End-to-end pipeline architecture
   - CRISP-DM phase labels
   - Data flow from sources through processing to models

4. **Schema Documentation** (`docs/PITCH_FEATURE_MART_SCHEMA.md`)
   - Complete table reference for all 7 tables
   - Column definitions and examples
   - Usage patterns (XGBoost, LSTM, feature selection)
   - Design principles (flexibility over optimization)
   - Data lineage visualization
   - Performance considerations

### Key Design Decisions

- **All 118 fields preserved** in base_features (not just "important" ones)
- **Metadata-driven queries** via feature_registry enable "SELECT features WHERE model = 'xgboost'"
- **Additive schema** - new features added without migrations
- **Versioned training sets** with SHA-256 data_hash for reproducibility
- **JSONB for sequences** enables variable-length LSTM windows

### Next Steps

1. Execute the SQL schema to create tables
2. Populate base_features from existing locations table
3. Build feature engineering pipeline
4. Create materialized training sets

Ready for code review and merge to develop.

**Comment by cbwinslow** [comment]
## 🎯 Epic #78 Progress Update - Pitch-Level Model Pipeline

**Date**: 2026-04-24  
**Status**: Phase 1-3 Complete ✅ | Phase 4 (Modeling) In Progress

---

### ✅ Completed Deliverables

#### 1. Data Population (7.66M Pitches)
| Table | Rows | Coverage |
|-------|------|----------|
| `features_pitch.base_features` | 7,661,992 | 2015-2025 seasons |
| `features_pitch.engineered_features` | 7,661,992 | All derived features |

**Data Quality**:
- All 118 Statcast fields preserved
- Tier 1 outcomes: S (69.9%), X (26.9%), B (3.2%)
- Tier 2 outcomes: 12 classes (Strikeout, Walk, Single, Double, Triple, HR, Out, HBP, Foul, Ball, Strike, Other)

#### 2. Feature Engineering (ALL Features - No Dropping)

| Category | Features | Research Backing |
|----------|----------|------------------|
| **Velocity** | `velocity_percentile`, `velocity_diff_from_avg` | SMU/CMU papers |
| **Strike Zone** | `distance_from_center`, `zone_region`, `is_in_zone`, `is_shadow`, `is_chase` | Zone judgment models |
| **Movement** | `horizontal_break`, `vertical_break`, `approach_angle`, `spin_efficiency`, `induced_vertical_break` | Pitch physics |
| **Game Context** | `score_diff`, `is_late_game`, `is_high_leverage` | Win probability |
| **Count** | `is_full_count`, `is_two_strike`, `is_three_ball` | Count dynamics |
| **Outcomes** | Tier 1: {S,B,X}, Tier 2: {12 classes} | Two-tier hierarchy |

#### 3. Scripts & Artifacts

**New Files**:
- `sql/features/004_alter_base_features_types.sql` - Schema fixes
- `sql/features/005_build_engineered_features.sql` - Feature engineering (22KB)
- `scripts/pitch_models/train_tier1_xgboost.py` - Tier-1 model training
- `docs/PITCH_MODEL_PROGRESS.md` - Progress documentation

**Scripts**:
- `populate_base_features.py` - Data migration with type casting
- `train_tier1_xgboost.py` - Stratified sampling, 21+ features

#### 4. Database Updates

**Schema Changes**:
- Extended `of_fielding_alignment` varchar(20) → varchar(50)
- Fixed `quality_flag` type issues
- Created indexes on `outcome_tier1`, `outcome_tier2`, `zone_region`

---

### 🔄 Phase 4: Modeling (In Progress)

#### Tier-1 XGBoost Baseline
- **Target**: >80% accuracy on S/B/X classification
- **Features**: 21+ engineered features (no dropping)
- **Training**: Stratified sampling, 200 trees, depth 6
- **Status**: Script ready, training in progress

#### Tier-2 Model (Next)
- **Target**: Fine-grained outcome prediction
- **Classes**: 12 outcome types
- **Features**: Same 21+ features

---

### 📊 Current Metrics

```
Valid Pitches for Training: 5,072,278 (excludes 'U' unknown)
Class Distribution (Tier 1):
  S (Strike):      3,546,222 (69.9%)
  X (Ball-in-Play): 1,364,716 (26.9%)
  B (Ball):         161,340  (3.2%)

Class Distribution (Tier 2 - Top 5):
  Ball:     2,587,846 (33.8%)
  Strike:   1,695,447 (22.1%)
  Foul:     1,412,452 (18.4%)
  Out:        885,555 (11.6%)
  Strikeout:  438,317  (5.7%)
```

---

### 🎯 Next Steps

1. **Complete Tier-1 Training** - Run full 5M pitch training
2. **Tier-2 Model** - Train 12-class classifier
3. **Evaluation** - Log Loss < 0.5, Brier < 0.25, Calibration < 0.05
4. **GitHub Actions** - Add model training CI/CD

---

### 📁 Files Modified/Created

```
sql/features/004_alter_base_features_types.sql     (NEW)
sql/features/005_build_engineered_features.sql     (NEW - 22KB)
scripts/pitch_models/train_tier1_xgboost.py        (NEW)
docs/PITCH_MODEL_PROGRESS.md                         (NEW)
```

### 🏷️ Labels
- `enhancement` `ml` `data-quality` `features`

### 🔗 Related
- Sub-Issue #79 (Schema Design) - ✅ Complete
- Parent Epic #78 - 🔄 Phase 4

---

**Commit**: `feat(pitch-models): Add engineered features and Tier-1 XGBoost training`  
**Author**: AI Agent  
**Epic**: #78

**Comment by cbwinslow** [comment]
Completed and superseded by v2.0 Production Hardening EPIC #117. All functionality delivered in the v2.0 release. Closing as completed.
